Feature add test documentation 411 (#424)
reactive-firewall/multicast: docs/Testing.md

> Multicast project testing code is under an MIT license like the Multicast module; however
> the project's Testing can use tools which are split between multiple licenses.
> While all the source-code is open-source, using some of the project test code is only possible
> in an environment where additional restrictions apply, due to third-party licensing which is
> incompatible if it were to be included with the rest of the project.
>
> Certain Continuous Integration/Continuous Deployment (CI/CD) tools may require GNU licensed
> software to operate effectively. These are _not_ included in the project.
> Environment Compatibility: Users must ensure that their testing environments are equipped
> with the necessary components and licenses to run the testing code. This is akin to a
> "batteries not included" disclaimer, indicating that additional setup may be required.
> AS-IS Disclaimer: Please note that the Multicast project is provided "AS-IS," and we do not
> guarantee compatibility or support for the testing code outside of the specified environments.
> Users are responsible for ensuring compliance with all applicable licenses and for setting up
> their environments accordingly.

> 

Author: 🙅🏻 The current wording is fine for now.

  * Also see [CEP-5](https://gist.github.com/reactive-firewall/3d2bd3cf37f87974df6f7bee31a05a89)

## Test Groups

### MATs

> [!IMPORTANT]
> Minimal acceptance tests are also automated as part of the CI processes, and run
> for each push to the repository. Minimal acceptance testing is run across multiple versions of
> Python to ensure stable behavior across a wide range of environments.

Minimal Acceptance Tests (MATs) are the bare minimum set of tests that must be successful for the
Multicast project to be considered acceptable.

#### CI/CD MATs

| Trigger | Workflow |
|---------|----------|
| `uses: .github/actions/run-minimal-acceptance-tests` | [.github/actions/run-minimal-acceptance-tests/action.yml](https://github.com/reactive-firewall/multicast/tree/HEAD/.github/actions/run-minimal-acceptance-tests/action.yml) |

#### Manual MATs

1. Clone a [development build](https://github.com/reactive-firewall/multicast/tree/patch-readme-docs?tab=readme-ov-file#developer-builds)

2. Invoke the MATs test target.

```bash
make test-mats
```

#### Categories

Currently, MATs are comprised of the test categories `bootstrap`, `basic`, `build`, `doctests`,
`say`, `hear`, and `usage`.

> 

Author: 🙅🏻 The current wording is preferred.

##### Coverage

The Multicast project uses CI/CD to ensure constant and comprehensive testing. `coverage` in the
`extra` group's context is referring to "extra test coverage".

> [!TIP]
> Contributors who are unsure of how to categorize their new test code, may fall back to the
> `coverage` category, so not to get bogged down before code review, where the categorization
> can better be discussed.

##### Linting

The Multicast project uses CI/CD to ensure all files, including code, configurations, and
documentation, are linted before publishing releases. This is automatically triggered by
pull requests during the project's release workflow, namely before merging anything to the
branch `master`, and again before releasing to `stable`.

> [!WARNING]
> Locally, linting is limited as many linters are copyleft restricted and can never be bundled with
> the rest of the MIT licensed code.

See implementation for more details.

> [!TIP]
> Want to check multicast python code style? Great! Try this:
>
> ```bash
> make clean ; # cleans up from any previous tests hopefully
> make test-style ; # runs the project's tests for style

> > make clean ; # cleans up for next test

Author: 🙅🏻 The current implementation is fine.

> ```

#### Logging and Coverage testing

By default, logging from the `multicast` module is ignored (there is a [default NullHandler](https://github.com/search?q=repo%3Areactive-firewall%2Fmulticast%20path%3Amulticast%2F*.py%20%22addHandler(logging.NullHandler())%22&type=code)
). The test-runner [`run_selective.py`](https://github.com/reactive-firewall/multicast/tree/HEAD/tests/run_selective.py)
inserts its own [special log handler](https://github.com/search?q=repo%3Areactive-firewall%2Fmulticast+path%3Atests%2F*.py+ColoredStreamHandler)
during testing to assist in debugging. Code coverage is only collected when invoking tests via the

> [`Makefile`](https://github.com/reactive-firewall/multicast/tree/HEAD/Makefile) targets that

Author: :thinking: the previous line is too long already, and moving the closing parentheses and period up too just makes it longer and weakens the intent of this being a single paragraph. :shrug: let's not get bogged down in the nitpicky here, at-least until a there is a more holistic solution to consider.

# Testing

Details about the Multicast project's own testing.

***

> [!CAUTION]
> Multicast project testing code is under an MIT license like the Multicast module; However
> the project's Testing can use tools which are split between multiple licenses.
> While all of the source-code is open-source, using some of the project test code is only possible
> in an environment where additional, restrictions apply, due to third party licensing which is
> incompatible if it were to be included with the rest of the project.

> >

Author: 🙅🏻 The current wording is fine for now.

> > [!CAUTION]
> > Multicast project testing code is under an MIT license like the Multicast module; However
> > the project's Testing can use tools which are split between multiple licenses.
> > While all of the source-code is open-source, using some of the project test code is only possible
> > in an environment where additional, restrictions apply, due to third party licensing which is
> > incompatible if it were to be included with the rest of the project.
> >
> > Certain Continuous Integration/Continuous Deployment (CI/CD) tools may require GNU licensed
> > software to operate effectively. These are _not_ included in the project.
> > Environment Compatibility: Users must ensure that their testing environments are equipped
> > with the necessary components and licenses to run the testing code. This is akin to a
> > "batteries not included" disclaimer, indicating that additional setup may be required.
> > AS-IS Disclaimer: Please note that the Multicast project is provided "AS-IS," and we do not
> > guarantee compatibility or support for the testing code outside of the specified environments.

Author: consider:
```suggestion
> [!CAUTION]
> Multicast project testing code is under an MIT license like the Multicast module; however
> the project's Testing can use tools which are split between multiple licenses.
> While all the source-code is open-source, using some of the project test code is only possible
> in an environment, where additional restrictions apply, due to third party licensing which is
> incompatible if it were to be included with the rest of the project.
>
> Certain Continuous Integration/Continuous Deployment (CI/CD) tools may require GNU licensed
> software to operate effectively. These are _not_ included in the project.
> Environment Compatibility: Users must ensure that their testing environments are equipped
> with the necessary components and licenses to run the testing code. This is akin to a
> "batteries not included" disclaimer, indicating that additional setup may be required.
> AS-IS Disclaimer: Please note that the Multicast project is provided "AS-IS," and we do not
> guarantee compatibility or support for the testing code outside of the specified environments.
```

does that work better? why, why not?

| `usage` | Tests focused on `multicast/__main__.py` and API use-cases | [test_usage.BasicIntegrationTestSuite](https://github.com/reactive-firewall/multicast/blob/HEAD/tests/test_usage.py) |

### Advanced Testing

> [!NOTE]
> # TL;DR - Contributor Focused
>
> The rest of this document is focused on topics that are intended only for those who wish to
> contribute (_or perhaps those few readers who are driven by an insatiable thirst for knowledge,
> and are here to delve into topics purely for the joy of learning_).

#### The Test Runner

> The test runner is an often obscure part of the Multicast project testing process, albeit an
> important part. In, general a test runner is exactly what it sounds like, the component responsible
> for running the various tests during the act of testing.

Author: Consider:
```suggestion
The test runner is an often obscure part of the Multicast project testing process, albeit an
important part. In general, a test runner is what it sounds like; the component responsible
for running the various tests during the act of testing.
```

What about a semi-colon?

#### Testing Performance

The Multicast Project includes some hooks for automating future performance testing.

> [!NOTE]
> Current performance testing is ad-hoc and focused on gathering more profiling data.

There is a separate test group intended for future performance testing aptly named `performance`.
Currently, the `performance` test group contains only empty categories named `scalability`,
`multi_sender` and `multi_receiver`.

| Category | _Intended_ Criteria | _TestSuites_ |
|----------|---------------------|--------------|
| `scalability` | Tests performance scaling | _Empty_ |
| `multi_sender` | Tests performance with multiple senders | _Empty_ |
| `multi_receiver` | Tests performance with multiple listeners | _Empty_ |

#### Testing even more

The Multicast Project includes some extra tests intended for additional, but non-acceptance,
testing (e.g., test for covering defensive or debugging code). These extra tests are organized
together under the aptly named group `extra`. The `extra` group contains the categories `coverage`,

> `linting`, and `security`.

Author: :no_good: in this case the distinction feels better emphasized with the commas, are they technically bad grammar syntax or just "unnecessary"?

the rest of the MIT licensed code.

> [!NOTE]
> See implementation for details.

> [!TIP]
> Want to check multicast python code style? Great! Try this:

```bash
make clean ; # cleans up from any previous tests hopefully
make test-style ; # runs the project's tests for style
make clean ; # cleans up for next test
```

#### Logging and Coverage testing

By default logging from the `multicast` module is ignored (there is a [default NULLHandeler](https://github.com/search?q=repo%3Areactive-firewall%2Fmulticast%20path%3Amulticast%2F*.py%20%22addHandler(logging.NullHandler())%22&type=code)
). The test-runner [`run_selective.py`](https://github.com/reactive-firewall/multicast/tree/HEAD/tests/run_selective.py)
inserts its own [special log handler](https://github.com/search?q=repo%3Areactive-firewall%2Fmulticast+path%3Atests%2F*.py+ColoredStreamHandler)
during testing to assist in debugging. Code coverage is only collected when invoking tests via the
[`Makefile`](https://github.com/reactive-firewall/multicast/tree/HEAD/Makefile) targets that
support it (and of-course only if already installed).

> [!TIP]

> > Code coverage can leave a lot of artifacts after testing, so it is a good idea to run the

Author: consider "numerous" instead:
```suggestion
> Code coverage can leave numerous artifacts after testing, so it is a good idea to run the
```

    CLI-->>user: starting
    CLI->>tests: Run loaded tests
    loop Testing
      tests->>tests: run test
      opt has output
        tests-->>Python: logs any multicast module output
        Python-->>logging: handle logging
        logging->>CLI: log messages
        CLI->>user: messages
      end
      alt passing
        tests-->>CLI: Successful test
        CLI-->>user: Reports test passed
      else skipping
        tests-->>CLI: will skip test
        CLI-->>user: Reports skipped
      else failing/error
        tests--xCLI: Test failure
        CLI-->>user: Reports Failed
      end
    end
    tests->>CLI: done
    CLI-->>user: Report Summary
    CLI-xPython: cleanup
    destroy logging
    Python--xlogging: cleanup
    Python--xtests: cleanup and exit
    tests->>Python: done
    Python->CLI: done
    CLI-->>user: ending
    CLI-xPython: exit
    Python->>CLI: exited
    opt if was via make
      CLI->>Make: exited
      alt with coverage.py installed
        Make-->COV: stop collecting
        destroy COV
        Make--xCOV: save and report coverage results
      end
      Make-->Make: do Make things
      Make->>user: done
    end
```

> #### As a Dependency, Testing

Author: 🙅🏻 The current wording is fine for now.
